flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp

```diff
@@ -0,0 +1,168 @@
+//===- GenericLoopConversion.cpp ------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "flang/Common/OpenMP-utils.h"
+#include "flang/Optimizer/Dialect/FIRDialect.h"
+#include "flang/Optimizer/HLFIR/HLFIRDialect.h"
+#include "flang/Optimizer/OpenMP/Passes.h"
+#include "flang/Semantics/symbol.h"
+
+#include "mlir/Dialect/ControlFlow/IR/ControlFlowOps.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/OpenMP/OpenMPDialect.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include <memory>
+
+namespace flangomp {
+#define GEN_PASS_DEF_GENERICLOOPCONVERSIONPASS
+#include "flang/Optimizer/OpenMP/Passes.h.inc"
+} // namespace flangomp
+
+namespace {
+
+/// A conversion pattern to handle various combined forms of `omp.loop`. For how
+/// combined/composite directive are handled see:
+/// https://discourse.llvm.org/t/rfc-representing-combined-composite-constructs-in-the-openmp-dialect/76986.
+class GenericLoopConversionPattern
+    : public mlir::OpConversionPattern<mlir::omp::LoopOp> {
+public:
+  enum class GenericLoopCombinedInfo {
+    None,
+    TargetTeamsLoop,
+    TargetParallelLoop
+  };
+
+  using mlir::OpConversionPattern<mlir::omp::LoopOp>::OpConversionPattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::omp::LoopOp loopOp, OpAdaptor adaptor,
+                  mlir::ConversionPatternRewriter &rewriter) const override {
+    assert(isLoopConversionSupported(loopOp));
+
+    rewriteToDistributeParallelDo(loopOp, rewriter);
+    rewriter.eraseOp(loopOp);
+    return mlir::success();
+  }
+
+  static bool isLoopConversionSupported(mlir::omp::LoopOp loopOp) {
```

Comment:
I did this intentionally to defer error reporting to `OpenMPToLLVMIRTranslation.cpp`. An unsupported loop will not result in a conversion failure. Rather the loop will be marked as "legal" and be allowed to continue down the pipeline. The advantage of doing this is to:
1. Use already existing error reporting in LLVM conversion.
2. During MLIR to MLIR conversion testing, we don't have to worry about handling errors from unsupported loop forms.

See also this [comment](https://github.com/llvm/llvm-project/pull/115443#discussion_r1846055696).

Reply:
Ah, I see. I thought this pass would fail because it emits an error. Then, it seems like we should emit a warning instead below.

The issue I see with that approach is that, whenever full support for the loop directive is implemented, no `omp.loop` ops are expected to reach translation to LLVM IR, since they would have been replaced by this pass. Hence, it seems to me that this is the right place to emit these errors now.

At the moment, translation to LLVM IR it will just abort saying "not yet implemented: omp.loop", which I think is not enough information. We could call `checkImplementationStatus` for it, and add cases for this operation to produce the proper per-clause TODO error messages, but I feel it wouldn't be the right place for it because the implementation is going to be in this pass.

Reply:
Added error messages and relevant tests.